### ISSUE TITLE
New/Edit themes form > Required fields indication

### ISSUE THREAD

**Issue by wion** [textpattern/textpattern#1433]
### Is your feature request related to a problem?

A suggestion

### What the hell is it?

In the New/Edit themes form, the first two fields are required but a user has to actually click the pophelp to learn that instead of there being obvious and conventional indicators right at surface of the form. I haven't checked if other such forms are like that.

Understandably you want theme designers to use as many metadata fields as possible, but, still. Little wins where you can?

**Comment by Bloke** [comment]
We should probably do this as a matter of course throughout. Indicate which are 'required' somehow. I always think it looks messy (because colour alone is not good enough), but if there's a way to do it somehow,then yes, it can save some frustration.

**Comment by wion** [comment]
One idea is simple ref-mark notation. Put an asterisk at end of the label (e.g. 'Theme name*') or on next to the text field itself, then add this at bottom of form: *Required. Done.

By the way, another thing I meant to bring up. When in context of a form like this, where the user came from the Themes panel, and the form header says **New theme** / **Edit theme**, it's somewhat redundant and monotonous to keep writing 'Theme' on every label. This is all you really need:

> ### New theme
>
> * Name*
> * Title *
> * Version
> * Description
> * Author
> * Website
>
> 
> *Required

That also eliminates the lopsided front-loading of the label, putting the true name word first, and capitalized.

**Comment by Bloke** [comment]
The tricky thing with using an asterisk or some ref notation is that it has to be part of the language strings. We can't (I don't think) just programmatically add it "after" since that breaks RTL languages.

Putting it in the language strings adds translation effort and, if we decide in future to flip the compulsory nature of a field, it has a knock-on effect top all language translators. I'd rather avoid that.

And yes I agree about overusing the word 'Theme'. As long as we don't reuse the string in other parts of the UI we can get away with shortening it. We should do that throughout as and when we find them. It might even save us a few strings, since things like 'Name' and 'Description' and 'Author' etc can go in the "admin-side" group and be reused across panels.

**Comment by wion** [comment]
What about putting the mark on the front of the text field? I see that done a lot too.

**Comment by wion** [comment]
Or making the _backgrounds_ of those two fields a slightly different tone (soft, soft blue?), then maybe no mark is needed at all.

**Comment by Bloke** [comment]
"In front" is subjective too. It's to the right in RTL languages. I'm not sure how browsers treat additional text we inject outside the strings. Although I suppose we can get away with it, because nobody will know if it was meant to be added 'after' the string or 'before' it.

```* テーマ名```

vs:

``` テーマ名 *```

Same animal, we're marking up the word and indicating its different nature.

With regards marking up the text box itself. We do have potential issues as I'm not sure if boxes themselves are subject to RTL - might only be only their content. Check by installing an RTL font and switching your admin side to it.

Shading would be my preferred option because it doesn't matter if you're red/green colourblind, you can see shades compared to white. BUT we need to consider:

* light mode and dark mode differences and how shades show up there.
* how to indicate - via some legend that takes up real-estate - what the shade means unless, as you say, the fact they're different is good enough indicator.

**Comment by wion** [comment]
Shading. No legend. Soft yellow in light mode. Dark blue or whatever in dark mode. I'm sure Dr. Phil can figure than one.

**Comment by Bloke** [comment]
Should just require the addition of some classes. Incidentally, we don't add the HTML `required` attribute right now because our input field UI library is lacking in this department. The new UI library (which I _might_ shoehorn into 4.8 even if it's unused by core in that version as there are only two or three more components to add) will support this natively and far more easily, allowing us to defer this to the browser too, preventing form submission unless the fields are filled in.

**Comment by phiw13** [comment]
> Shading. No legend. Soft yellow in light mode. Dark blue or whatever in dark mode. I'm sure Dr. Phil can figure than one.

No. or use that as a secondary addition to a real indicator.

> Should just require the addition of some classes. 

Yes, this. On the ` <label >` or the parent `<div />`; theme authors can then add some indicator using the `::before` and `::after` pseudo classes. Even programatically injecting a character/symbol inside or before/after the label ought to work well.

> I'm not sure if boxes themselves are subject to RTL

They are, as it set on the root element.  ` html dir=rtl`

The current situation in Hive is indeed not good. the required fields get some coloured border on `:focus`. It is a dark red, which to my eyes looks barely different from the normal black border. And the cultural differences regarding the meanings of the those colours.

**Comment by Bloke** [comment]
> The current situation in Hive is indeed not good. the required fields get some coloured border on `:focus`. It is a dark red, which to my eyes looks barely different from the normal black border. And the cultural differences regarding the meanings of the those colours.

Actually, I lied. We do use `required` in a few places. And when we do, the _browser_ adds the dark red border on focus, not us. At least, my Firefox does.

**Comment by Bloke** [comment]
> [add a class] on the ` <label >` or the parent `<div />`; theme authors can then add some indicator using the `::before` and `::after` pseudo classes.

Winner.

**Comment by phiw13** [comment]
> Actually, I lied. 

Nope. the CSS below is what does the thing.

```
input:focus:invalid, select:focus:invalid, textarea:focus:invalid {
border-color: #9d261d;
box-shadow: none;
}

**Comment by wion** [comment]
> No. or use that as a secondary addition to a real indicator.

Better than nothing at all, was the point.

But if a real indicator can be swung, swing it.

**Comment by philwareham** [comment]
We can put an indicator after the label text (still in the label tag) if necessary. It doesn’t break RTL at all as it’s still rendered after the text - regardless of if that text is running right or left.

It should be relatively simple to do as I think I’ve already put `required` attributes on any required fields. Do we can just search those out and add extra code.

**Comment by Bloke** [comment]
If it's relatively easy to do, sure, by all means either do it or let me know what markup to add. If we can add it as phiw13 suggests using ::before / ::after in CSS, then that's also great. I'm easy, whichever is the best approach.

**Comment by philwareham** [comment]
For reference see:

https://www.w3.org/TR/WCAG20-TECHS/H90.html
https://www.w3.org/TR/WCAG20-TECHS/G83.html

### TLDR

Add this text somewhere...

    Fields marked <abbr class="required" title="Required">*</abbr> are required.

(we already have translations of that text as part of our public theme, although it'd have to be extracted from a longer text string sentence and put in a new string.

Add this **within** the <label> tag of any required fields (after the label text):

    <abbr class="required" title="Required">*</abbr>

Then we comply with WGAG 2.0 H90. To comply with G83 also, we would need a bit more code (and I'm not sure we would really need to, as we already have `required` attributes on form fields which tells the screen reader if the field is required/invalid on submit).

**Comment by philwareham** [comment]
For discussion: https://design-patterns.textpattern.com/mockups/secondary/presentation-skin_edit.html

Note that I have put the indicator *before* the pophelp so it is announced in a prompt manner (rather than after an announcement of the pophelp link).

**Comment by Bloke** [comment]
Niiiice!

**Comment by phiw13** [comment]
> For discussion: https://design-patterns.textpattern.com/mockups/secondary/presentation-skin_edit.html

Yes. The indicator should certainly belong before the pophelp (I tend to forget those help thingies as I have them hidden on my work installs.

I wouldn’t worry too much about G83 as all supported browser support the `required` attribute (the browser UI takes over here), screen reader support is good I’m told, and there is some post submission routine in TXP, although it does not bring you back to the form in case of failure.

**Comment by philwareham** [comment]
OK, when I get a spare moment I'll code this in (prior to 4.8 release). Thanks all.

**Comment by philwareham** [comment]
@bloke I think I'll need some help with the PHP around this request.

So we need to put a reusable bit of text on the labels of fields that are required. Something like...

```
$required_label = sp.tag('*', 'b', array(
        'class' => 'required',
        'title' => gTxt('required'),
    ));
```

I have already added code to allow `required` attributes on [`function fInput()`](https://github.com/textpattern/textpattern/blob/dev/textpattern/lib/txplib_forms.php#L316) and [`function text_area()`](https://github.com/textpattern/textpattern/blob/dev/textpattern/lib/txplib_forms.php#L578) and it would be handy if we could use that to determine whether to add the extra text to the label tag or not.

However, I'm aware that `function inputLabel()` generates the label and structure around it - and I'm unsure how/if that function can determine if any child `fInput` or `text_area` functions have required=true.

If you can provide code or advise me, that would be very handy.

**Comment by Bloke** [comment]
Hmmm, I'm not sure either. Let me mull it over today and I'll see if anything can be done.

**Comment by philwareham** [comment]
Thanks. This also needs to be a consideration in the new widgets too.

**Comment by Bloke** [comment]
Yeah, beard-stroking in progress :)

**Comment by Bloke** [comment]
We can (sort of) do this via a little hack to simply `strpos()` the `$input` string to look for `required="required"`. If it finds it, we can add the asterisk markup. Wonderful:

![Screen Shot 2019-12-03 at 12 35 18](https://user-images.githubusercontent.com/21252/70051662-6bd5b980-15c9-11ea-8852-1c46e6e891ce.png)

However, that broad-brush approach also causes this:

![Screen Shot 2019-12-03 at 12 38 44](https://user-images.githubusercontent.com/21252/70051882-e4d51100-15c9-11ea-839e-fcfaaf46f5d2.png)

Clearly, although the browsed file is required, it's not required unless you're using it. Hmmm.

We also have issues of where to display the 'Fields marked <abbr class="required" title="Required">*</abbr> are required.' text on panels such as the Pages panel:

![Screen Shot 2019-12-03 at 12 38 10](https://user-images.githubusercontent.com/21252/70051944-08985700-15ca-11ea-8678-be5f08300bca.png)

Ideas  many of the above?

[Incidentally, in the new UI library, a simple method can be employed to check if a widget has the `required` attribute set or not, as we have access to the individually embedded objects. That's way more robust than checking for a string in a string.]

**Comment by Bloke** [comment]
Incidentally, this also filters through to plugins so they automatically get the asterisk on any fields that are nominated as `required` - as long as they've used `inputLabel()` of course. So the same problem applies on file input fields. If indeed it is a problem.

**Comment by Bloke** [comment]
Here's the patch if anybody wants to play or improve it:

``` php
    $requiredLabel = (strpos($input, 'required="required"') === false)
        ? ''
        : sp.tag('*', 'b', array(
            'class' => 'required',
            'title' => gTxt('required'),
        ));

    if ($label !== '') {
        $labelContent = tag(gTxt($label).$requiredLabel.popHelp($help[0]), 'label', array('for' => $name)).$tools;
    } else {
        $labelContent = gTxt($name).$requiredLabel.popHelp($help[0]).$tools;
    }

```

It goes after `$inlineHelp = (isset($help[1])) ? $help[1] : '';` in function `inputLabel()` in 'txplib_html.php', replacing the `if ($label !== '')` block after it.

**Comment by philwareham** [comment]
Hmmm sorry, maybe we are overthinking this.

I have now slightly changed the handling of invalid fields in the CSS. Any required fields that are not filled in (apart from file type fields) are now indicated by a red border regardless of whether they are in focus or not. I have also brightened the error red colour we use to make it stand out a little better.

Whilst this isn't as blinding obvious as the required indicators and supporting description, it is probably good enough (and screen readers announce the field is required anyway).

Otherwise we scatter a lot of extra clutter into forms that becomes more of a distraction than benefit.

**Comment by Bloke** [comment]
Works for me. Less visual clutter too. If people are happy with this - both those that do and don't use assistive technologies - then great.

**Comment by bloatware** [comment]
We could also automatically add `required` placeholder if none is set to required fields for colour-blind people?

**Comment by philwareham** [comment]
Sure, I'm fine with that suggestion. We have the `gTxt` available. Can you do?

**Comment by bloatware** [comment]
Sure, just need to know which functions are concerned: `fInput()`, `text_area()`, what else?

**Comment by philwareham** [comment]
@bloatware just those two, yes

**Comment by philwareham** [comment]
@bloatware although probably worth mentioning that `input` tag `placerholder` attributes only affect types: `password`, `search`, `tel`, `text`, and `url`.

**Comment by bloatware** [comment]
Ok, done it, but plugin upload textarea is not generated via `text_area()` neither really required any more.

**Comment by philwareham** [comment]
@bloatware thanks. I'm happy with this.

Yes it seems `text_area()` is hardly used in the code, rather we are using HTML. I think this is primarily because that function lacks allowance things such as `class` and `dir` attributes. When we move to widgets we can adjust that.

I'm happy to close this issue if everyone is happy with the result.

**Comment by Bloke** [comment]
Fine here. Is the email field on the _New user_ panel not included because we don't target `type="email"` ?

**Comment by philwareham** [comment]
Oh, good point. Yes we should include `email` too. Seems that is error/omission on MDN documentation. Fixed in our code now.

**Comment by bloatware** [comment]
> 
> 
> Yes it seems `text_area()` is hardly used in the code, rather we are using HTML. I think this is primarily because that function lacks allowance things such as `class` and `dir` attributes. When we move to widgets we can adjust that.

You actually can pass it any array of attributes as the first argument (idem for `fInput()`).

**Comment by Bloke** [comment]
Yeah, until you made those changes, `text_area()` was seriously lacking in functionality so we rarely used it. Can't wait to ditch this parameter overloading madness for something cleaner!

**Comment by petecooper** [comment]
![Screen Shot 2019-12-03 at 19 42 48](https://user-images.githubusercontent.com/226578/70083901-33ec6780-1605-11ea-9d09-0670463280f6.png)

I've got to say, that's rather nicely done.

**Comment by phiw13** [comment]
Feeling a bit sad if the independent indicator goes away.

Based on the couple of screenshots that @Bloke posted above, I don’t see any issue with the clutter. It adds no more visual bagage than the help buttons do now.

I’m no fan of making the `required` fields stand out the way it is done now. And it is on the edge of being (very) hard to distinguish for people with colour deficiencies. The `placeholder` attribute is not a bad addition though. But put it between ()?

Screenschot with SimDaltonism set to Deuteranopia
<img width="572" alt="Screen Shot 2019-12-04 at 8 07 37" src="https://user-images.githubusercontent.com/8944809/70098453-51eeb200-166f-11ea-867a-33500ff792c2.png">

**Comment by Bloke** [comment]
Now we've fixed the 'required' status of the file input fields so the indicators wouldn't appear there, I'm not averse to using the actual asterisk indicator. I like the placeholder (and yes if it was in parentheses or something, fine). The red shows up okay for me, but I can appreciate how it might not for others.

If we were to adopt asterisks - either now or (better, because it's more robust) using the new UI library when it's folded into core - the only thing that concerns me is where we would put the explanatory text about what the asterisks mean. It's fine on 'Edit' steps with a conventional form, but for things like the Pages and Forms panels, I don't know where it would fit comfortably.

Is the notion of using an asterisk indicator to denote 'required' so universally accepted that we don't need explanatory text nearby?

I am a big fan of how DigitalOcean approach this, btw. They use taller input boxes and use (presumably) JavaScript to pop messages up above what you type. These work better than placeholders, imo because you don't even need labels. The labels are part of the input field and remain visible even after you type.

I'm not saying we should adopt this today, I just like the approach. Here's an example. This is what happens when you land on the page. Focused on the Name box.

![Screen Shot 2019-12-03 at 23 40 59](https://user-images.githubusercontent.com/21252/70099523-ef71c380-1626-11ea-91dc-f8c0195a343a.png)

Note the field label is inside the field, making it obvious what the field is for. Just like placeholder but more in-your-face. Note also that there's an asterisk on the right and no explanatory text in sight.

If you leave the box without filling it out, the input placeholder 'Enter name' becomes a tiny bit smaller (it's a fast animation) and the explanatory text appears above it in red, as does the asterisk and the box border:

![Screen Shot 2019-12-03 at 23 41 25](https://user-images.githubusercontent.com/21252/70099642-50010080-1627-11ea-9a1b-c3f3baa4cf94.png)

When you start typing in the field, the field label moves (again, an animation) to where the error message was, so you know what the field represents even if you come back to it in future. With a regular placeholder, you need a separate label because the placeholder is removed when you start typing. With this approach, everything is all in one place:

![Screen Shot 2019-12-03 at 23 41 55](https://user-images.githubusercontent.com/21252/70099722-93f40580-1627-11ea-9c59-b314845e2bca.png)

Same deal with dropdowns. If you try to leave the field, the error text slides into view above the placeholder.

As I say, not advocating we use it, but it ticks a lot of boxes for me. Is that approach good for assistive technologies?

Side note: I have absolutely no idea how they render it, because it sure as hell isn't HTML. All they have in the page source is a bare skeleton of a few container divs and then the forms seem to be injected into them from some JS library. Maybe React, I dunno. If we were to ever do this, I'd be keen to do it with as much HTML and `data-` elements as possible, using JS only to augment the UX.

But anyway. Food for thought... or not.

Here's a [30-second video (SFW) of the above in action](https://stefdawson.com/downloads/support/DO-InputError.mov).

**Comment by phiw13** [comment]
> If we were to adopt asterisks - either now or (better, because it's more robust) using the new UI library when it's folded into core - the only thing that concerns me is where we would put the explanatory text about what the asterisks mean. It's fine on 'Edit' steps with a conventional form, but for things like the Pages and Forms panels, I don't know where it would fit comfortably.

Oh, sure, I’m not saying this must be done in a great hurry. The current situation has existed for a while without killing anyone (I hope ?). And if it is more easy to do with your magic library, sure. And the addition of the placeholder _is_ helpful. Yeah, the explanation text, that is indeed a problem on the Pages/… and Write panels. No answer on that atm.

As for the DO effects. Seems a bit popular on techies websites. For a short form (your example), maybe, maybe. For longer / complex ones I’m not convinced.

PS - I you right-click on the element and select ‘Inspect Element’ (Safari), does it tell you a fuller story?

**Comment by phiw13** [comment]
I guess something that bothers me a little (in Hive) is using the `:invalid` pseudo-class to style the field when empty, even on page load. Semantically they are not in an invalid state (yet).

When many many years ago Gecko implemented this `required` feature with the associated invalid/valid pseudo-classes, there was a huge outcry. Things went along the “I haven’t interacted with the form yet, why does it show the field as invalid?” and some variations.
The Gecko people then came up with some nifty heuristics, the fields would not be marked invalid on load, only after (failed) submit, and the `-moz-ui-invalid` pseudo-class was invented. If the authors used the `:valid` and `:invalid` ones, nothing changed.

Alternative selectors to style those fields: `[required]:empty {}` and `[required]:placeholder-shown {}`.
(the latter won’t do anything in IE 11 and Edge pre-betrayal)

And yeah, if the user empty the field and puts nothing back in then the state _is_ invalid. But you can’t detect and then style all those different heuristics in CSS.

**Comment by philwareham** [comment]
FWIW `[required]:empty {}` and `[required]:placeholder-shown {}` don't work how you expect them to.

I've done a partial revert of the invalid pseudo class so it only shows on focus (and additional in Firefox after submission when not focussed). That'll have to do for now.

If/when `:user-invalid` [becomes available](https://drafts.csswg.org/selectors-4/#user-pseudos) (the official version of `:-moz-ui-invalid`) we can use that to improve things too.

Closing.

**Comment by wion** [comment]
The 4.8.0-dev demo still shows a missing 'required' indicator in the 'Theme title' field of the **New theme** form. Not sure that's intentional.
